### PR TITLE
Retry certain requests on failure

### LINKED ISSUE #898: Retry proxied requests when network level errors occur
Need to implement the same kind of retry logic as this lib:

https://github.com/FGRibreau/node-request-retry

['ECONNRESET', 'ENOTFOUND', 'ESOCKETTIMEDOUT', 'ETIMEDOUT', 'ECONNREFUSED', 'EHOSTUNREACH', 'EPIPE', 'EAI_AGAIN']

TODO: research `EPIPE` we may not be able to retry from that.

### LINKED ISSUE #1013: proxy mishandles network errors
- Operating System: ubuntu 14.04
- Cypress Version: 1.1.3
- Browser Version: chrome 60

### Is this a Feature or Bug?
bug

### Current behavior:
Connection refused errors, that should fail in a fraction of a second, are instead returned as timeouts, after 30s.

### Desired behavior:
Return error 'connection refused' on connection refused, at the time the connection is refused.

### How to reproduce:
In the test app, make an ajax request to port that isn't listening. Outside cypress, it should fail immediately with connection refused. In cypress, it will fail after half a minute with timeout.

#### Test code:

```js

```

### LINKED ISSUE #4013: Retry automatically on `cy.visit`, `cy.request`, server check
We should automatically retry some operations for the user, to mimic the behavior of a normal Chrome action.

The following code was used to test the scenarios against a real Chrome instance:

```js
const http = require('http')
const debug = require('debug')('test')

const immediateServer = http.createServer((req, res) => {
    debug(`received ${req.method} ${req.url}`)
    req.socket.destroy()
})

immediateServer.listen(12345)

const afterHeadersServer = http.createServer((req, res) => {
    debug(`received ${req.method} ${req.url}`)
    res.writeHead(200)
    res.write('')
    setTimeout(() => req.socket.destroy(), 2000)
})

afterHeadersServer.listen(12346)

const midBodyServer = http.createServer((req, res) => {
    debug(`received ${req.method} ${req.url}`)
    res.writeHead(200)
    res.write('<html>')
    setTimeout(() => req.socket.destroy(), 2000)
})

midBodyServer.listen(12347)
```

### Page visits

* If the request fails before receiving data, wait 5s before retrying once. It looks like Chrome retries with a backoff of 30 seconds each time, but we likely don't want to wait this long - just try 3 times with the user's `commandTimeout` between each
```
  test received GET / +0ms
  test received GET / +34ms
  test received GET / +5s
  test received GET / +30s
  test received GET / +1m
```
* If the request fails after receiving headers but before data is sent (second case), the behavior is the same as above, so we should do the same:
```
  test received GET / +0ms
  test received GET / +37ms
  test received GET / +5s
  test received GET / +30s
  test received GET / +1m
```
* If the request ends mid-body, Chrome doesn't retry or consider it an error:
```
  test received GET / +0ms
```

### XHRs

* In all 3 cases, Chrome does not automatically retry a failed GET XHR:
```
  test received GET / +0ms
```
* Chrome will also not automatically retry a failed POST XHR:
```
  test received POST / +0ms
```

### Conclusion

* Implement automatic retries on `cy.visit()` network errors before body is received, try 3 times with `commandTimeout` wait each time
* Same for `cy.request()`
* Retry baseurl check 3x, with backoff of 1s, 2s, 2s
* Pass through network errors so Chrome's normal retry behavior kicks in too

### PR DESCRIPTION
**Pull request by flotwig**
Closes #4013
Closes #898
Closes #1013

- [x] Retry cy.visit(), cy.request(), and all requests that go through Cypress under the hood
	- If `options.retryOnNetworkFailure` is `true` (default) or it's a Cypress proxy request, `ECONNREFUSED`, `ECONNRESET`, `EPIPE`, `EHOSTUNREACH`, and `EAI_AGAIN` errors will be retried
	- If `options.retryOnStatusCodeFailure` is `true`, requests will be retried on non-successful status codes
	- Retries will happen 4 times at increasing intervals: 0s, 1s, 2s, 2s
	- `ECONNREFUSED` retries and proxied retries happen at shorter intervals: 0s, 100ms, 200ms, 200ms
	- Proxied request bodies will be buffered in memory as they are streamed in so that we can re-send them in the event of a network error
- [x] Set `{ timeout: null }` on proxied requests so the only timeout is the eventual OS timeout, since Chrome won't timeout
- [x] Retry `https-proxy` connections in proxied mode and in direct mode using same rules
- [x] Retry the `baseUrl` check in `run` mode at 0s, 1s, 2s, 2s intervals, then fail if it is unsuccessful
- [x] Update documentation: cypress-io/cypress-documentation#1648

### CODE REVIEW

**Comment by flotwig** [comment]
New baseurl check output in run mode:

```
Cypress could not verify that the server set as your `baseUrl` is running: http://never-gonna-exist.invalid

Your tests likely make requests to this `baseUrl` and these tests will fail if you don't boot your server.

We will retry 3 more times in 1 second...

Cypress could not verify that the server set as your `baseUrl` is running: http://never-gonna-exist.invalid

Your tests likely make requests to this `baseUrl` and these tests will fail if you don't boot your server.

We will retry 2 more times in 2 seconds...

Cypress could not verify that the server set as your `baseUrl` is running: http://never-gonna-exist.invalid

Your tests likely make requests to this `baseUrl` and these tests will fail if you don't boot your server.

We will retry 1 more time in 2 seconds...
Cypress could not verify that the server set as your `baseUrl` is running:

  > http://never-gonna-exist.invalid

Your tests likely make requests to this `baseUrl` and these tests will fail if you don't boot your server.

Please start this server and then run Cypress again.
```

**Comment by flotwig** [comment]
Experiments with Chrome's behavior in proxy mode:

## Timeouts

### Proxy accepts connection but sends no data back, leaves socket open

#### First attempt (took so long that Node `http.Server` timed out, not the browser!)
* Wireshark filter: `ip.addr == 127.0.0.1 && tcp.port == 1234 && http.request.uri contains "http://example.com"`
* Server code: `require('http').createServer((req, res) => console.log(req.method, req.url)).listen(1234)`
* Chrome launch: `chromium-browser --proxy-server=http://127.0.0.1:1234`
* Q: Does Chrome eventually give up and retry after sending the request to the proxy and waiting forever?
    * A: In this experiment, Chrome hits the proxy every 120 seconds:
image
    * *However*, it's actually Node that's ending the conversation, not the browser in this scenario - this is a screenshot of the total conversation for the first retry. Notice that the `FIN` comes from `1234`, not from `47294` (Chrome's [ephemeral port](https://en.wikipedia.org/wiki/Ephemeral_port)). This was causing an `ERR_EMPTY_RESPONSE` in Chrome, not a timeout. Kinda weird because the node.js docs state that by default, sockets have [no timeout](https://nodejs.org/api/net.html#net_socket_settimeout_timeout_callback) and [no keepalive](https://nodejs.org/api/net.html#net_socket_setkeepalive_enable_initialdelay) - maybe sockets from `http.Server` are different. image
    * This is just the normal retrying on `ERR_EMPTY_RESPONSE` behavior which we know about already

#### With .setKeepAlive(true), .setTimeout(0)
* Wireshark filter: `ip.addr == 127.0.0.1 && tcp.port == 1234 && http.request.uri contains "http://example.com"`
* Server code: 
    ```js
    require('http').createServer((req, res) => {
        console.log(req.method, req.url);
        res.connection.setKeepAlive(true);
        res.connection.setTimeout(0);
    }).listen(1234)
    ```
* Chrome launch: `chromium-browser --proxy-server=http://127.0.0.1:1234`
* Q: Does Chrome eventually give up and retry after sending the request to the proxy and waiting forever?
    * Nooooope. Here you can see Chrome happily waiting for 10 minutes while it bats TCP keep-alives back and forth for 10 minutes before I stopped recording.  Pasted image

## Browser close behavior

* Wireshark filter: `ip.addr == 127.0.0.1 && tcp.port == 1234 && http.request.uri contains "http://example.com"`
* Server code to hang a socket open forever: 
    ```js
    require('http').createServer((req, res) => {
        console.log(req.method, req.url);
        res.connection.setKeepAlive(true);
        res.connection.setTimeout(0);
    }).listen(1234)
    ```
* Chrome launch: `chromium-browser --proxy-server=http://127.0.0.1:1234`

### When requests are still pending and browser is closed cleanly
* Q: What happens to the outstanding requests?
    * The last three packets in the following capture occurred after exiting. You can see that the TCP stream is closed cleanly. Pasted image

### When requests are still pending and browser is `kill -9`'d
* Command used: `killall --signal 9 chromium-browser`
* Q: What happens to the outstanding requests?
    * They are still closed cleanly after the browser is force killed: Pasted image

File: packages/driver/src/cy/commands/navigation.coffee [hunk 1]
```diff
@@ -492,6 +494,9 @@ module.exports = (Commands, Cypress, cy, state, config) ->
         onLoad: ->
       })
 
+      if options.retryOnStatusCodeFailure and not options.failOnStatusCode
```
Comment (flotwig) [inline packages/driver/src/cy/commands/navigation.coffee]: add a test for this

File: packages/driver/src/cy/commands/request.coffee [hunk 2]
```diff
@@ -81,15 +84,17 @@ module.exports = (Commands, Cypress, cy, state, config) ->
...
-        failOnStatusCode: true
+        log: true
       })
 
       ## if timeout is not supplied, use the configured default
       options.timeout ||= config("responseTimeout")
 
       options.method = options.method.toUpperCase()
 
+      if options.retryOnStatusCodeFailure and not options.failOnStatusCode
```
Comment (flotwig) [inline packages/driver/src/cy/commands/request.coffee]: add a test for this

File: packages/server/lib/project.coffee [hunk 3]
```diff
@@ -71,6 +72,8 @@ class Project extends EE
 
       @memoryCheck = setInterval(logMemory, 1000)
 
+    @onWarning = options.onWarning
```
Comment (flotwig) [inline packages/server/lib/project.coffee]: this is getting passed in multiple places, stick with one

**Comment by bahmutov** [comment]
Building binary on Mac does show a problem with new network package
```
NODE_ENV is undefined
npm install --only=production --quiet
in folder: /Users/gleb/git/cypress/dist/darwin/packages/coffee
npm WARN deprecated coffee-script@1.12.7: CoffeeScript on NPM has moved to "coffeescript" (no hyphen)
packages/runner/lib -> /Users/gleb/git/cypress/dist/darwin/packages/runner/lib
packages/network/index -> /Users/gleb/git/cypress/dist/darwin/packages/network/index
packages/desktop-gui/lib -> /Users/gleb/git/cypress/dist/darwin/packages/desktop-gui/lib
added 1 package from 1 contributor and audited 1 package in 0.721s
found 0 vulnerabilities

removing devDependencies from /Users/gleb/git/cypress/dist/darwin/packages/driver/package.json
installing /Users/gleb/git/cypress/dist/darwin/packages/driver
NODE_ENV is undefined
npm install --only=production --quiet
in folder: /Users/gleb/git/cypress/dist/darwin/packages/driver
packages/network/index -> /Users/gleb/git/cypress/dist/darwin/packages/network/index
packages/network/index -> /Users/gleb/git/cypress/dist/darwin/packages/network/index
packages/network/index -> /Users/gleb/git/cypress/dist/darwin/packages/network/index
Unhandled rejection Error: operation timed out after 4011 ms, 5 tries with error: ENOENT: no such file or directory, stat 'packages/network/index'
    at /Users/gleb/git/cypress/node_modules/bluebird-retry/lib/bluebird-retry.js:109:35
```

**Comment by brian-mann** [comment]
@bahmutov yup this is what i was explaining on the phone that I knew would be a problem

**Comment by YOU54F** [comment]
Hey @flotwig / @brian-mann , I just pulled this commit into my fork whilst testing microsoft edge and I can't load any browsers now in gui or cli mode, I get the error in the terminal output below. 

Just removed pulling master into my fork (this [pr](https://github.com/YOU54F/cypress/pull/1) and it has resolved it for all browsers, just a friendly fyi 👍 

I am running a mbp 2018 with mojave 10.14.4

```
➜  cypress git:(feat/msEdge) ✗ npm run start
Alias tip: npmR start

> cypress@3.2.0 prestart /Users/you54f/dev/saftest/githubrepos/forks/cypress
> npm run check-deps-pre


> cypress@3.2.0 check-deps-pre /Users/you54f/dev/saftest/githubrepos/forks/cypress
> node ./scripts/check-deps.js --verbose --prescript


> cypress@3.2.0 start /Users/you54f/dev/saftest/githubrepos/forks/cypress
> node ./cli/bin/cypress open --dev --global

GET /__/ 200 23.792 ms - -
TypeError: connect.createRetryingSocket is not a function
    at Server._makeConnection (/Users/you54f/dev/saftest/githubrepos/forks/cypress/packages/https-proxy/lib/server.coffee:144:20)
    at Server._makeDirectConnection (/Users/you54f/dev/saftest/githubrepos/forks/cypress/packages/https-proxy/lib/server.coffee:103:6)
    at Server._onFirstHeadBytes (/Users/you54f/dev/saftest/githubrepos/forks/cypress/packages/https-proxy/lib/server.coffee:71:17)
    at Socket.<anonymous> (/Users/you54f/dev/saftest/githubrepos/forks/cypress/packages/https-proxy/lib/server.coffee:62:8)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:594:20)
```

**Comment by flotwig** [comment]
@YOU54F try deleting all of the .js files from packages/network/lib, you have an old version of a TS file transpiled

**Comment by YOU54F** [comment]
Spot on @flotwig , sorted it, thank you for that!

**Pull request merged**